src/exo/cursors.py

```diff
@@ -0,0 +1,591 @@
+            yield from []
+        elif isinstance(n, LoopIR.USub):
+            yield from self._children_from_attrs(n, 'arg')
+        elif isinstance(n, LoopIR.BinOp):
+            yield from self._children_from_attrs(n, 'lhs', 'rhs')
+        elif isinstance(n, LoopIR.BuiltIn):
+            yield from self._children_from_attrs(n, 'args')
+
+    def _children_from_attrs(self, n, *args) -> Iterable[Node]:
+        for attr in args:
+            children = getattr(n, attr)
+            if isinstance(children, list):
+                for i in range(len(children)):
+                    yield self.child(attr, i)
+            else:
+                yield self.child(attr, None)
+
+    # ------------------------------------------------------------------------ #
+    # Navigation (block selectors)
+    # ------------------------------------------------------------------------ #
+
+    def body(self) -> Block:
+        return self._attr_block('body')
+
+    def orelse(self) -> Block:
+        return self._attr_block('orelse')
+
+    def _attr_block(self, attr: str):
+        stmts = getattr(self._node(), attr)
+        assert isinstance(stmts, list)
+        return Block(self._proc, self._path + [(attr, range(len(stmts)))])
+
+    # ------------------------------------------------------------------------ #
+    # Conversions
+    # ------------------------------------------------------------------------ #
+
+    def as_block(self) -> Block:
+        attr, i = self._path[-1]
+        if i is None:
+            raise InvalidCursorError('node is not inside a block')
+        return Block(self._proc, self._path[:-1] + [(attr, range(i, i + 1))])
+
+    # ------------------------------------------------------------------------ #
+    # AST mutation
+    # ------------------------------------------------------------------------ #
+
+    def _replace(self, ast):
+        """
+        This is an UNSAFE internal function for replacing a node in an AST with
+        either a list of statements or another single node and providing a
+        forwarding function as collateral. It is meant to be package-private,
+        not class-private, so it may be called from other internal classes and
+        modules, but not from end-user code.
+        """
+        if self._path[-1][1] is not None:
+            # noinspection PyProtectedMember
+            # delegate block replacement to the Block class
+            return self.as_block()._replace(ast)
+
+        # replacing a single expression, or something not in a block
+        def update(node):
+            attr, _ = self._path[-1]
+            return node.update(**{attr: ast})
+
+        from .API import Procedure
+        p = Procedure(self._rewrite_node(update))
+
+        return p, self._forward_replace(weakref.ref(p))
+
+    def _forward_replace(self, new_proc):
+        assert self._path[-1][1] is None
+
+        def fwd_node(_):
+            raise InvalidCursorError('cannot forward replaced nodes')
```

Comment:
As mentioned before, this is not necessarily true when sub-trees are involved.  This might also not be true when you have a node in the middle of a block being replaced.  A good sanity check is that if the `_replace` method above delegates to the `Block` form of `_replace`, does the forwarding policy similarly delegate?